### PR TITLE
Update Windows support to Cluster

### PR DESCRIPTION
**Pull request by gowripadmajas**
* The field "aws_auth_node_iam_role_arns_windows" would enable role mapping in the aws-auth for Windows worker.
* ConfigMap: "amazon-vpc-cni": enables Windows support to the cluster.

Screenshot 2022-07-05 at 12 06 14 PM

### CODE REVIEW

File: modules/self_managed_nodes/win_cm_vpc_cni.tf [hunk 1]
```diff
@@ -0,0 +1,22 @@
+locals {
+  aws_auth_configmap_data = {
+    enable-windows-ipam = "true"
+  }
+}
+
+resource "kubernetes_config_map" "amazon-vpc-cni" {
```
Comment (lawliet89) [inline modules/self_managed_nodes/win_cm_vpc_cni.tf:7]:
```suggestion
resource "kubernetes_config_map" "amazon_vpc_cni" {
```

Snake case for names.

File: modules/self_managed_nodes/win_cm_vpc_cni.tf [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    namespace = "kube-system"
+  }
+
+  data = local.aws_auth_configmap_data
+
+  lifecycle {
+    # We are ignoring the data here since we will manage it with the resource below
+    # This is only intended to be used in scenarios where the configmap does not exist
+    ignore_changes = [data]
+  }
```
Comment (lawliet89) [inline modules/self_managed_nodes/win_cm_vpc_cni.tf:21]: This should not be here, right?

File: modules/self_managed_nodes/win_cm_vpc_cni.tf [hunk 3]
```diff
@@ -0,0 +1,22 @@
+locals {
+  aws_auth_configmap_data = {
+    enable-windows-ipam = "true"
+  }
+}
+
+resource "kubernetes_config_map" "amazon-vpc-cni" {
+  count = var.create_aws_auth_configmap ? 1 : 0
```
Comment (lawliet89) [inline modules/self_managed_nodes/win_cm_vpc_cni.tf:8]: Able to add a variable to turn on support for Windows nodes? Then we only set the workers IAM role and create this resource when windows support is needed.

Comment (gowripadmajas) [reply modules/self_managed_nodes/win_cm_vpc_cni.tf]: Updated as suggested.

File: variables.tf [hunk 4]
```diff
@@ -65,6 +65,12 @@ variable "manage_aws_auth_configmap" {
   default     = true
 }
 
+variable "enable_cluster_windows_support" {
+  description = "Determines whether to create the amazon-vpc-cni configmap and windows worker roles into aws-auth."
+  type        = bool
+  default     = false
+}
+
```
Comment (lawliet89) [inline variables.tf:73]: This is not necessary at the root module, right? And in any case, it's not passed down to the child module.

File: modules/self_managed_nodes/variables.tf [hunk 5]
```diff
@@ -32,6 +32,13 @@ variable "worker_security_group_id" {
 # Refer to https://github.com/terraform-aws-modules/terraform-aws-eks/tree/master/modules/self-managed-node-group
 # for the parameters supported. See README for more information.
 ####################################
+
+variable "enable_cluster_windows_support" {
+  description = "Determines whether to create the amazon-vpc-cni configmap and windows worker roles into aws-auth."
+  type        = bool
+  default     = false
+}
+
```
Comment (lawliet89) [inline modules/self_managed_nodes/variables.tf:41]: Do update the README for this module too.

File: node_groups.tf [hunk 6]
```diff
@@ -132,3 +132,24 @@ resource "aws_cloudwatch_event_target" "node_termination_handler_spot" {
   rule      = aws_cloudwatch_event_rule.node_termination_handler_spot.name
   arn       = module.node_termination_handler_sqs.sqs_queue_arn
 }
+
+################################################
+# Below resources are to Enable Windows support to cluster.
```
Comment (lawliet89) [inline node_groups.tf:137]:
```suggestion
# Resources below are to enable Windows Support for Cluster.
```

**Review by lawliet89 — APPROVED** [review]

**Comment by lawliet89** [comment]
Good to go

**Pull request merged**